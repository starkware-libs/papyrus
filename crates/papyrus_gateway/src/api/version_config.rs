#[derive(Eq, PartialEq, Hash)]
/// Labels the jsonRPC versions we have such that one and only one version is Latest,
/// there can be multiple versions that are supported (latest is implicitly supported), and there
/// can be multiple versions that are deprecated.
/// Latest -> method exposed via the http path "/" and "" (e.g. http://host:port/)
/// Supported -> method exposed via the http path "/version_id" (e.g. http://host:port/V0_3_0)
/// Deprecated -> method not exposed.
pub enum VersionState {
    // TODO: nevo - remove the dead_code attribute once other versions are implemented - hides
    // "Supported" and "Deprecated" not constructed error
    Latest,
    #[allow(dead_code)]
    Supported,
    #[allow(dead_code)]
    Deprecated,
}

pub const VERSION_0_3_0: &str = "V0_3_0";
pub const VERSION_CONFIG: &[(&str, VersionState)] = &[(VERSION_0_3_0, VersionState::Latest)];
