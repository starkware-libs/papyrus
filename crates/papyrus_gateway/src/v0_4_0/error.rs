use jsonrpsee::types::ErrorObjectOwned;

#[derive(Clone, Debug)]
pub enum JsonRpcError {
    FailedToReceiveTransaction,
    ContractNotFound,
    InvalidTransactionHash,
    InvalidBlockHash,
    BlockNotFound,
    InvalidTransactionIndex,
    ClassHashNotFound,
    TransactionHashNotFound,
    PageSizeTooBig,
    NoBlocks,
    InvalidContinuationToken,
    TooManyKeysInFilter,
    ContractError,
    // TODO(dvir): delete this when start support pending blocks.
    PendingBlocksNotSupported,
    ClassAlreadyDeclared,
    InvalidTransactionNonce,
    InsufficientMaxFee,
    InsufficientAccountBalance,
    ValidationFailure,
    CompilationFailed,
    ContractClassSizeIsTooLarge,
    NonAccount,
    DuplicateTx,
    CompiledClassHashMismatch,
    UnsupportedTxVersion,
    UnsupportedContractClassVersion,
    UnexpectedError(String),
}

impl From<JsonRpcError> for ErrorObjectOwned {
    fn from(err: JsonRpcError) -> Self {
        match err {
            JsonRpcError::FailedToReceiveTransaction => {
                ErrorObjectOwned::owned(1, "Failed to write transaction", None::<()>)
            }
            JsonRpcError::ContractNotFound => {
                ErrorObjectOwned::owned(20, "Contract not found", None::<()>)
            }
            JsonRpcError::InvalidTransactionHash => {
                ErrorObjectOwned::owned(25, "Invalid transaction hash", None::<()>)
            }
            JsonRpcError::InvalidBlockHash => {
                ErrorObjectOwned::owned(26, "Invalid block hash", None::<()>)
            }
            JsonRpcError::BlockNotFound => {
                ErrorObjectOwned::owned(24, "Block not found", None::<()>)
            }
            JsonRpcError::InvalidTransactionIndex => {
                ErrorObjectOwned::owned(27, "Invalid transaction index in a block", None::<()>)
            }
            JsonRpcError::ClassHashNotFound => {
                ErrorObjectOwned::owned(28, "Class hash not found", None::<()>)
            }
            JsonRpcError::TransactionHashNotFound => {
                ErrorObjectOwned::owned(29, "Transaction hash not found", None::<()>)
            }
            JsonRpcError::PageSizeTooBig => {
                ErrorObjectOwned::owned(31, "Requested page size is too big", None::<()>)
            }
            JsonRpcError::NoBlocks => {
                ErrorObjectOwned::owned(32, "There are no blocks", None::<()>)
            }
            JsonRpcError::InvalidContinuationToken => ErrorObjectOwned::owned(
                33,
                "The supplied continuation token is invalid or unknown",
                None::<()>,
            ),
            JsonRpcError::TooManyKeysInFilter => {
                ErrorObjectOwned::owned(34, "Too many keys provided in a filter", None::<()>)
            }
            JsonRpcError::ContractError => {
                ErrorObjectOwned::owned(40, "Contract error", None::<()>)
            }
            JsonRpcError::PendingBlocksNotSupported => ErrorObjectOwned::owned(
                41,
                "Currently, Papyrus doesn't support pending blocks.",
                None::<()>,
            ),
            JsonRpcError::ClassAlreadyDeclared => {
                ErrorObjectOwned::owned(51, "Class already declared", None::<()>)
            }
            JsonRpcError::InvalidTransactionNonce => {
                ErrorObjectOwned::owned(52, "Invalid transaction nonce", None::<()>)
            }
            JsonRpcError::InsufficientMaxFee => ErrorObjectOwned::owned(
                53,
                "Max fee is smaller than the minimal transaction cost (validation plus fee \
                 transfer)",
                None::<()>,
            ),
            JsonRpcError::InsufficientAccountBalance => ErrorObjectOwned::owned(
                54,
                "Account balance is smaller than the transaction's max_fee",
                None::<()>,
            ),
            JsonRpcError::ValidationFailure => {
                ErrorObjectOwned::owned(55, "Account validation failed", None::<()>)
            }
            JsonRpcError::CompilationFailed => {
                ErrorObjectOwned::owned(56, "Compilation failed", None::<()>)
            }
            JsonRpcError::ContractClassSizeIsTooLarge => {
                ErrorObjectOwned::owned(57, "Contract class size it too large", None::<()>)
            }
            JsonRpcError::NonAccount => {
                ErrorObjectOwned::owned(58, "Sender address in not an account contract", None::<()>)
            }
            JsonRpcError::DuplicateTx => ErrorObjectOwned::owned(
                59,
                "A transaction with the same hash already exists in the mempool",
                None::<()>,
            ),
            JsonRpcError::CompiledClassHashMismatch => ErrorObjectOwned::owned(
                60,
                "the compiled class hash did not match the one supplied in the transaction",
                None::<()>,
            ),
            JsonRpcError::UnsupportedTxVersion => {
                ErrorObjectOwned::owned(61, "the transaction version is not supported", None::<()>)
            }
            JsonRpcError::UnsupportedContractClassVersion => ErrorObjectOwned::owned(
                62,
                "the contract class version is not supported",
                None::<()>,
            ),
            JsonRpcError::UnexpectedError(data) => {
                ErrorObjectOwned::owned(63, "An unexpected error occured", Some(data))
            }
        }
    }
}
