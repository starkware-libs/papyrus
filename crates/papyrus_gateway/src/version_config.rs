#[derive(Eq, PartialEq, Hash)]
/// Labels the jsonRPC versions we have such that there can be multiple versions that are supported,
/// and there can be multiple versions that are deprecated.
/// Supported -> method exposed via the http path "/version_id" (e.g. http://host:port/V0_3_0)
/// Deprecated -> method not exposed.
#[derive(Clone, Copy, Debug)]
pub enum VersionState {
    // TODO: nevo - remove the dead_code attribute once other versions are implemented - hides
    // "Supported" and "Deprecated" not constructed error
    Supported,
    #[allow(dead_code)]
    Deprecated,
}

/// latest version must be set as supported
pub const VERSION_CONFIG: &[(&str, VersionState)] =
    &[(VERSION_0_3_0, VersionState::Supported), (VERSION_0_4_0, VersionState::Supported)];
pub const VERSION_0_3_0: &str = "V0_3_0";
pub const VERSION_0_4_0: &str = "V0_4_0";
