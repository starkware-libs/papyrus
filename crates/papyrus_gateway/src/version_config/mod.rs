pub mod version_config;
#[cfg(test)]
mod version_config_test;
