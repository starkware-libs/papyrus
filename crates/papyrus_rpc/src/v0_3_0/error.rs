use jsonrpsee::types::ErrorObjectOwned;

#[derive(thiserror::Error, Clone, Copy, Debug)]
pub enum JsonRpcError {
    #[error("Contract not found")]
    ContractNotFound = 20,
    #[error("Block not found")]
    BlockNotFound = 24,
    #[error("Transaction hash not found")]
    TransactionHashNotFound = 25,
    #[error("Invalid transaction index in a block")]
    InvalidTransactionIndex = 27,
    #[error("Class hash not found")]
    ClassHashNotFound = 28,
    #[error("Transaction reverted")]
    TransactionReverted = 29,
    #[error("Requested page size is too big")]
    PageSizeTooBig = 31,
    #[error("There are no blocks")]
    NoBlocks = 32,
    #[error("The supplied continuation token is invalid or unknown")]
    InvalidContinuationToken = 33,
    #[error("Too many keys provided in a filter")]
    TooManyKeysInFilter = 34,
    #[error("Contract error")]
    ContractError = 40,
}

impl From<JsonRpcError> for ErrorObjectOwned {
    fn from(err: JsonRpcError) -> Self {
        ErrorObjectOwned::owned(err as i32, err.to_string(), None::<()>)
    }
}
