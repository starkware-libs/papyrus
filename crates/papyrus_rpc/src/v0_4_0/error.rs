use jsonrpsee::types::ErrorObjectOwned;

#[derive(Clone, Debug)]
pub struct JsonRpcError {
    pub code: i32,
    pub message: &'static str,
    pub data: Option<String>,
}

// TODO(yair): Remove allow(dead_code) once all errors are used.
#[allow(dead_code)]
pub const FAILED_TO_RECEIVE_TRANSACTION: JsonRpcError =
    JsonRpcError { code: 1, message: "Failed to write transaction", data: None };

pub const CONTRACT_NOT_FOUND: JsonRpcError =
    JsonRpcError { code: 20, message: "Contract not found", data: None };

pub const INVALID_TRANSACTION_HASH: JsonRpcError =
    JsonRpcError { code: 25, message: "Invalid transaction hash", data: None };

pub const INVALID_BLOCK_HASH: JsonRpcError =
    JsonRpcError { code: 26, message: "Invalid block hash", data: None };

pub const BLOCK_NOT_FOUND: JsonRpcError =
    JsonRpcError { code: 24, message: "Block not found", data: None };

pub const INVALID_TRANSACTION_INDEX: JsonRpcError =
    JsonRpcError { code: 27, message: "Invalid transaction index in a block", data: None };

pub const CLASS_HASH_NOT_FOUND: JsonRpcError =
    JsonRpcError { code: 28, message: "Class hash not found", data: None };

pub const TRANSACTION_HASH_NOT_FOUND: JsonRpcError =
    JsonRpcError { code: 29, message: "Transaction hash not found", data: None };

pub const PAGE_SIZE_TOO_BIG: JsonRpcError =
    JsonRpcError { code: 31, message: "Requested page size is too big", data: None };

pub const NO_BLOCKS: JsonRpcError =
    JsonRpcError { code: 32, message: "There are no blocks", data: None };

pub const INVALID_CONTINUATION_TOKEN: JsonRpcError = JsonRpcError {
    code: 33,
    message: "The supplied continuation token is invalid or unknown",
    data: None,
};

pub const TOO_MANY_KEYS_IN_FILTER: JsonRpcError =
    JsonRpcError { code: 34, message: "Too many keys provided in a filter", data: None };

pub const SERVER_NOT_SYNCED: JsonRpcError =
    JsonRpcError { code: 35, message: "Server is not synchronized", data: None };

pub const CONTRACT_ERROR: JsonRpcError =
    JsonRpcError { code: 40, message: "Contract error", data: None };

pub const CLASS_ALREADY_DECLARED: JsonRpcError =
    JsonRpcError { code: 51, message: "Class already declared", data: None };

pub const INVALID_TRANSACTION_NONCE: JsonRpcError =
    JsonRpcError { code: 52, message: "Invalid transaction nonce", data: None };

pub const INSUFFICIENT_MAX_FEE: JsonRpcError = JsonRpcError {
    code: 53,
    message: "Max fee is smaller than the minimal transaction cost (validation plus fee transfer)",
    data: None,
};

pub const INSUFFICIENT_ACCOUNT_BALANCE: JsonRpcError = JsonRpcError {
    code: 54,
    message: "Account balance is smaller than the transaction's max_fee",
    data: None,
};

pub const VALIDATION_FAILURE: JsonRpcError =
    JsonRpcError { code: 55, message: "Account validation failed", data: None };

pub const COMPILATION_FAILED: JsonRpcError =
    JsonRpcError { code: 56, message: "Compilation failed", data: None };

pub const CONTRACT_CLASS_SIZE_IS_TOO_LARGE: JsonRpcError =
    JsonRpcError { code: 57, message: "Contract class size it too large", data: None };

pub const NON_ACCOUNT: JsonRpcError =
    JsonRpcError { code: 58, message: "Sender address in not an account contract", data: None };

pub const DUPLICATE_TX: JsonRpcError = JsonRpcError {
    code: 59,
    message: "A transaction with the same hash already exists in the mempool",
    data: None,
};

pub const COMPILED_CLASS_HASH_MISMATCH: JsonRpcError = JsonRpcError {
    code: 60,
    message: "the compiled class hash did not match the one supplied in the transaction",
    data: None,
};

pub const UNSUPPORTED_TX_VERSION: JsonRpcError =
    JsonRpcError { code: 61, message: "the transaction version is not supported", data: None };

pub const UNSUPPORTED_CONTRACT_CLASS_VERSION: JsonRpcError =
    JsonRpcError { code: 62, message: "the contract class version is not supported", data: None };

pub fn unexpected_error(data: String) -> JsonRpcError {
    JsonRpcError { code: 64, message: "An unexpected error occured", data: Some(data) }
}

impl From<JsonRpcError> for ErrorObjectOwned {
    fn from(err: JsonRpcError) -> Self {
        ErrorObjectOwned::owned(err.code, err.message, err.data)
    }
}
