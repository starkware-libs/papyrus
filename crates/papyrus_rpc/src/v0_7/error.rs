use jsonrpsee::types::ErrorObjectOwned;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug)]
pub struct JsonRpcError<T: Serialize> {
    pub code: i32,
    pub message: &'static str,
    pub data: Option<T>,
}

// TODO(yair): Remove allow(dead_code) once all errors are used.
#[allow(dead_code)]
pub const FAILED_TO_RECEIVE_TRANSACTION: JsonRpcError<String> =
    JsonRpcError { code: 1, message: "Failed to write transaction", data: None };

pub const CONTRACT_NOT_FOUND: JsonRpcError<String> =
    JsonRpcError { code: 20, message: "Contract not found", data: None };

pub const INVALID_TRANSACTION_HASH: JsonRpcError<String> =
    JsonRpcError { code: 25, message: "Invalid transaction hash", data: None };

// TODO(shahak): Remove allow(dead_code) once all errors are used.
#[allow(dead_code)]
pub const INVALID_BLOCK_HASH: JsonRpcError<String> =
    JsonRpcError { code: 26, message: "Invalid block hash", data: None };

pub const BLOCK_NOT_FOUND: JsonRpcError<String> =
    JsonRpcError { code: 24, message: "Block not found", data: None };

pub const INVALID_TRANSACTION_INDEX: JsonRpcError<String> =
    JsonRpcError { code: 27, message: "Invalid transaction index in a block", data: None };

pub const CLASS_HASH_NOT_FOUND: JsonRpcError<String> =
    JsonRpcError { code: 28, message: "Class hash not found", data: None };

pub const TRANSACTION_HASH_NOT_FOUND: JsonRpcError<String> =
    JsonRpcError { code: 29, message: "Transaction hash not found", data: None };

pub const PAGE_SIZE_TOO_BIG: JsonRpcError<String> =
    JsonRpcError { code: 31, message: "Requested page size is too big", data: None };

pub const NO_BLOCKS: JsonRpcError<String> =
    JsonRpcError { code: 32, message: "There are no blocks", data: None };

pub const INVALID_CONTINUATION_TOKEN: JsonRpcError<String> = JsonRpcError {
    code: 33,
    message: "The supplied continuation token is invalid or unknown",
    data: None,
};

pub const TOO_MANY_KEYS_IN_FILTER: JsonRpcError<String> =
    JsonRpcError { code: 34, message: "Too many keys provided in a filter", data: None };

#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
pub struct ContractError {
    pub revert_error: String,
}

impl From<ContractError> for JsonRpcError<ContractError> {
    fn from(contract_error: ContractError) -> Self {
        Self { code: 40, message: "Contract error", data: Some(contract_error) }
    }
}
#[derive(Debug, Clone, Deserialize, Serialize, Eq, PartialEq)]
pub struct TransactionExecutionError {
    pub transaction_index: usize,
    pub execution_error: String,
}

impl From<TransactionExecutionError> for JsonRpcError<TransactionExecutionError> {
    fn from(tx_execution_error: TransactionExecutionError) -> Self {
        Self { code: 41, message: "Transaction execution error", data: Some(tx_execution_error) }
    }
}
pub const CLASS_ALREADY_DECLARED: JsonRpcError<String> =
    JsonRpcError { code: 51, message: "Class already declared", data: None };

pub const INVALID_TRANSACTION_NONCE: JsonRpcError<String> =
    JsonRpcError { code: 52, message: "Invalid transaction nonce", data: None };

pub const INSUFFICIENT_MAX_FEE: JsonRpcError<String> = JsonRpcError {
    code: 53,
    message: "Max fee is smaller than the minimal transaction cost (validation plus fee transfer)",
    data: None,
};

pub const INSUFFICIENT_ACCOUNT_BALANCE: JsonRpcError<String> = JsonRpcError {
    code: 54,
    message: "Account balance is smaller than the transaction's max_fee",
    data: None,
};

pub fn validation_failure(data: String) -> JsonRpcError<String> {
    JsonRpcError { code: 55, message: "Account validation failed", data: Some(data) }
}

pub const COMPILATION_FAILED: JsonRpcError<String> =
    JsonRpcError { code: 56, message: "Compilation failed", data: None };

pub const CONTRACT_CLASS_SIZE_IS_TOO_LARGE: JsonRpcError<String> =
    JsonRpcError { code: 57, message: "Contract class size it too large", data: None };

pub const NON_ACCOUNT: JsonRpcError<String> =
    JsonRpcError { code: 58, message: "Sender address in not an account contract", data: None };

pub const DUPLICATE_TX: JsonRpcError<String> = JsonRpcError {
    code: 59,
    message: "A transaction with the same hash already exists in the mempool",
    data: None,
};

pub const COMPILED_CLASS_HASH_MISMATCH: JsonRpcError<String> = JsonRpcError {
    code: 60,
    message: "the compiled class hash did not match the one supplied in the transaction",
    data: None,
};

pub const UNSUPPORTED_TX_VERSION: JsonRpcError<String> =
    JsonRpcError { code: 61, message: "the transaction version is not supported", data: None };

pub const UNSUPPORTED_CONTRACT_CLASS_VERSION: JsonRpcError<String> =
    JsonRpcError { code: 62, message: "the contract class version is not supported", data: None };

pub fn unexpected_error(data: String) -> JsonRpcError<String> {
    JsonRpcError { code: 63, message: "An unexpected error occurred", data: Some(data) }
}

impl<T: Serialize> From<JsonRpcError<T>> for ErrorObjectOwned {
    fn from(err: JsonRpcError<T>) -> Self {
        ErrorObjectOwned::owned(err.code, err.message, err.data)
    }
}
