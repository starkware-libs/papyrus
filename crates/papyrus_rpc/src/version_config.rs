#[cfg(test)]
#[path = "version_config_test.rs"]
mod version_config_test;

use std::fmt;

pub const VERSION_PATTERN: &str = "[Vv][0-9]+_[0-9]+(_[0-9]+)?";

#[derive(Eq, PartialEq, Hash)]
/// Labels the jsonRPC versions we have such that there can be multiple versions that are supported,
/// and there can be multiple versions that are deprecated.
/// Supported -> method exposed via the http path "/version_id" (e.g. http://host:port/V0_3_0)
/// Deprecated -> method not exposed.
#[derive(Clone, Copy, Debug)]
pub enum VersionState {
    // TODO: nevo - remove the dead_code attribute once other versions are implemented - hides
    // "Supported" and "Deprecated" not constructed error
    Supported,
    #[allow(dead_code)]
    Deprecated,
}

#[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
pub struct VersionId {
    // TODO(yair): change to enum so that the match in get_methods_from_supported_apis can be
    // exhaustive.
    pub name: &'static str,
    pub patch: u8,
}

impl fmt::Display for VersionId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}_{}", self.name, self.patch)
    }
}

/// latest version must be set as supported
pub const VERSION_CONFIG: &[(VersionId, VersionState)] = &[
    (VERSION_0_4, VersionState::Supported),
    (VERSION_0_5, VersionState::Supported),
    (VERSION_0_6, VersionState::Supported),
    (VERSION_0_7, VersionState::Supported),
];
pub const VERSION_0_4: VersionId = VersionId { name: "V0_4", patch: 0 };
pub const VERSION_0_5: VersionId = VersionId { name: "V0_5", patch: 1 };
pub const VERSION_0_6: VersionId = VersionId { name: "V0_6", patch: 0 };
pub const VERSION_0_7: VersionId = VersionId { name: "V0_7", patch: 0 };
