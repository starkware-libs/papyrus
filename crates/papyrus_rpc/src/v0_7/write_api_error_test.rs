use std::collections::BTreeMap;

use enum_iterator::all;
use jsonrpsee::types::ErrorObjectOwned;
use starknet_client::starknet_error::{KnownStarknetErrorCode, StarknetError, StarknetErrorCode};

use super::super::error::JsonRpcError;
use super::{
    starknet_error_to_declare_error,
    starknet_error_to_deploy_account_error,
    starknet_error_to_invoke_error,
};
use crate::test_utils::{get_starknet_spec_api_schema_for_method_errors, SpecFile};
use crate::version_config::VERSION_0_7 as Version;

const MESSAGE: &str = "message";
const UNKNOWN_CODE: &str = "code";

fn starknet_errors() -> impl Iterator<Item = StarknetError> {
    all::<KnownStarknetErrorCode>()
        .map(StarknetErrorCode::KnownErrorCode)
        .chain([StarknetErrorCode::UnknownErrorCode(UNKNOWN_CODE.to_owned())])
        .map(|error_code| StarknetError { code: error_code, message: MESSAGE.to_owned() })
}

fn test_error_from_conversion_fits_rpc<F: Fn(StarknetError) -> JsonRpcError<String>>(
    f: F,
    spec_method: &str,
) {
    let schema = get_starknet_spec_api_schema_for_method_errors(
        &[(SpecFile::WriteApi, &[spec_method])],
        &Version,
    );
    for starknet_error in starknet_errors() {
        // Converting into ErrorObjectOwned since it has serialization.
        let rpc_error: ErrorObjectOwned = f(starknet_error).into();
        let mut json_value = serde_json::to_value(rpc_error).unwrap();
        json_value.as_object_mut().unwrap().retain(|_, v| !v.is_null());
        assert!(schema.is_valid(&json_value));
    }
}

#[test]
fn starknet_error_to_invoke_error_result_fits_specs() {
    test_error_from_conversion_fits_rpc(
        starknet_error_to_invoke_error,
        "starknet_addInvokeTransaction",
    );
}

#[test]
fn starknet_error_to_declare_error_result_fits_specs() {
    test_error_from_conversion_fits_rpc(
        starknet_error_to_declare_error,
        "starknet_addDeclareTransaction",
    );
}

#[test]
fn starknet_error_to_deploy_account_error_result_fits_specs() {
    test_error_from_conversion_fits_rpc(
        starknet_error_to_deploy_account_error,
        "starknet_addDeployAccountTransaction",
    );
}

fn get_conversion_snapshot<F: Fn(StarknetError) -> JsonRpcError<String>>(
    f: F,
) -> serde_json::Value {
    // Using BTreeMap to keep the keys sorted.
    let mut rpc_error_code_to_errors = BTreeMap::new();
    for starknet_error in starknet_errors() {
        let rpc_error: ErrorObjectOwned = f(starknet_error.clone()).into();
        rpc_error_code_to_errors.insert(rpc_error.code(), (rpc_error, starknet_error));
    }
    serde_json::to_value(rpc_error_code_to_errors).unwrap()
}

#[test]
fn starknet_error_to_invoke_error_snapshot() {
    insta::assert_json_snapshot!(get_conversion_snapshot(starknet_error_to_invoke_error));
}

#[test]
fn starknet_error_to_declare_error_snapshot() {
    insta::assert_json_snapshot!(get_conversion_snapshot(starknet_error_to_declare_error));
}

#[test]
fn starknet_error_to_deploy_account_error_snapshot() {
    insta::assert_json_snapshot!(get_conversion_snapshot(starknet_error_to_deploy_account_error));
}
