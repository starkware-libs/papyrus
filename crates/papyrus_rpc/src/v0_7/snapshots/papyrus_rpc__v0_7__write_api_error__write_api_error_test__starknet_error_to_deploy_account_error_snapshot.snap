---
source: crates/papyrus_rpc/src/v0_7/write_api_error_test.rs
expression: get_conversion_snapshot(starknet_error_to_deploy_account_error)
---
{
  "28": [
    {
      "code": {
        "$serde_json::private::Number": "28"
      },
      "message": "Class hash not found"
    },
    {
      "code": "StarknetErrorCode.UNDECLARED_CLASS",
      "message": "message"
    }
  ],
  "52": [
    {
      "code": {
        "$serde_json::private::Number": "52"
      },
      "message": "Invalid transaction nonce"
    },
    {
      "code": "StarknetErrorCode.INVALID_TRANSACTION_NONCE",
      "message": "message"
    }
  ],
  "53": [
    {
      "code": {
        "$serde_json::private::Number": "53"
      },
      "message": "Max fee is smaller than the minimal transaction cost (validation plus fee transfer)"
    },
    {
      "code": "StarknetErrorCode.INSUFFICIENT_MAX_FEE",
      "message": "message"
    }
  ],
  "54": [
    {
      "code": {
        "$serde_json::private::Number": "54"
      },
      "message": "Account balance is smaller than the transaction's max_fee"
    },
    {
      "code": "StarknetErrorCode.INSUFFICIENT_ACCOUNT_BALANCE",
      "message": "message"
    }
  ],
  "55": [
    {
      "code": {
        "$serde_json::private::Number": "55"
      },
      "data": "message",
      "message": "Account validation failed"
    },
    {
      "code": "StarknetErrorCode.VALIDATE_FAILURE",
      "message": "message"
    }
  ],
  "58": [
    {
      "code": {
        "$serde_json::private::Number": "58"
      },
      "message": "Sender address in not an account contract"
    },
    {
      "code": "StarknetErrorCode.ENTRY_POINT_NOT_FOUND_IN_CONTRACT",
      "message": "message"
    }
  ],
  "59": [
    {
      "code": {
        "$serde_json::private::Number": "59"
      },
      "message": "A transaction with the same hash already exists in the mempool"
    },
    {
      "code": "StarknetErrorCode.DUPLICATED_TRANSACTION",
      "message": "message"
    }
  ],
  "61": [
    {
      "code": {
        "$serde_json::private::Number": "61"
      },
      "message": "the transaction version is not supported"
    },
    {
      "code": "StarknetErrorCode.INVALID_TRANSACTION_VERSION",
      "message": "message"
    }
  ],
  "63": [
    {
      "code": {
        "$serde_json::private::Number": "63"
      },
      "data": "message",
      "message": "An unexpected error occurred"
    },
    {
      "code": "code",
      "message": "message"
    }
  ]
}
